### PR TITLE
contributor.name to_fedora mapping fix for structuredValue

### LINKED ISSUE #1218: Cocina map contributor namePart in a structuredValue without a type
It should be able to round trip the following XML as found in stage on druid:jh029dg0916

```xml
  <name type="personal" usage="primary">
    <namePart>Sarmiento, Domingo Faustino</namePart>
    <namePart type="date">1811-1888</namePart>
  </name>
  <name type="personal">
    <namePart>Rojas, Ricardo</namePart>
    <namePart type="date">1882-1957</namePart>
  </name>
```

Currently it's has nil name parts:
```xml
  <name type="personal" usage="primary">
    <namePart/>
  </name>
  <name type="personal">
    <namePart/>
  </name>
```

### PR DESCRIPTION
**Pull request by ndushay**
## Why was this change made?

Fixes #1218

- corrects to_fedora mapping for contributor.name.namePart when it is a structured value
- bonus:   added to_fedora and from_fedora specs for "multiple names, one primary" - started with this thinking it would trip the problem, but it turned out NOT to trip the problem in #1218, AND it turned out to be already implemented.

## How was this change tested?

functional tests were written with the same data as was in the record in ticket #1218 -- the from_fedora as well as the to_fedora.

### CODE REVIEW

**Comment by ndushay** [comment]
File sul-dlss-labs/cocina-descriptive-metadata/issues/102 so Arcadia can map an example akin to this particular data.

File: app/services/cocina/to_fedora/descriptive/contributor.rb [hunk 1]
```diff
@@ -59,9 +59,9 @@ def write_roles_for(contributor)
                 attributes[:type] = 'code'
                 value = role.code
               end
-              attributes[:valueURI] = role.uri
-              attributes[:authority] = role.source&.code
-              attributes[:authorityURI] = role.source&.uri
+              attributes[:valueURI] = role.uri if role.uri
+              attributes[:authority] = role.source&.code if role.source&.code
```
Comment (jcoyne) [inline app/services/cocina/to_fedora/descriptive/contributor.rb:63]: I don't think you need the `&.` twice here. It's only needed in the conditional clause.

File: app/services/cocina/to_fedora/descriptive/contributor.rb [hunk 2]
```diff
@@ -59,9 +59,9 @@ def write_roles_for(contributor)
...
+              attributes[:authorityURI] = role.source&.uri if role.source&.uri
```
Comment (jcoyne) [inline app/services/cocina/to_fedora/descriptive/contributor.rb:64]: same here.

File: app/services/cocina/to_fedora/descriptive/contributor.rb [hunk 3]
```diff
@@ -21,7 +21,7 @@ def initialize(xml:, contributors:)
 
         def write
           Array(contributors).each_with_index do |contributor, _alt_rep_group|
-            if contributor.name.count == 1
+            if contributor.name.count == 1 && contributor.name.first.structuredValue.blank?
```
Comment (jcoyne) [inline app/services/cocina/to_fedora/descriptive/contributor.rb:24]: Do we need to test for blank here, or is it really a check for nil?

File: app/services/cocina/to_fedora/descriptive/contributor.rb [hunk 4]
```diff
@@ -70,13 +70,13 @@ def write_roles_for(contributor)
...
-              Array(name.structuredValue).each do |part|
-                xml.namePart part.value, type: NAME_PART.fetch(part.type)
+              Array(name.structuredValue).each do |name_part|
+                write_structured_name_part(name_part)
               end
 
               xml.displayForm name.value if name.type == 'display'
             end
-            contributor.note.each do |note|
+            contributor.note&.each do |note|
```
Comment (jcoyne) [inline app/services/cocina/to_fedora/descriptive/contributor.rb:79]: How about using `Array(contributor.note)` as we do on 73 for consistency.

File: app/services/cocina/to_fedora/descriptive/contributor.rb [hunk 5]
```diff
@@ -87,11 +87,17 @@ def write_complex(contributor)
               end
             end
 
-            contributor.identifier.each do |ident|
+            contributor.identifier&.each do |ident|
```
Comment (jcoyne) [inline app/services/cocina/to_fedora/descriptive/contributor.rb:90]: How about using `Array(contributor.note)` as we do on 73 for consistency.

**Review by jcoyne — APPROVED** [review]
I'm so excited for this one. I think it'll really make a big dent in the accuracy of the mapping. However, we should hold until the mapping stuff is sorted out.

File: app/services/cocina/to_fedora/descriptive/contributor.rb (hunk 3 above)
Comment (ndushay) [reply app/services/cocina/to_fedora/descriptive/contributor.rb:24]: nil is correct, you're right.

**Comment by ndushay** [comment]
@jcoyne -- took all your suggestions - thanks!    In reality, this isn't a new mapping because I didn't touch from_fedora mappings.  So I'll leave it to you to merge it or change it to a draft.

**Comment by jcoyne** [comment]
@ndushay there are a few conflicts to fix.

**Comment by ndushay** [comment]
@jcoyne rebased.  (For the second time)

**Review by jcoyne — APPROVED** [review]
